docs/DesignDocs/Protected_Code_Loader_Proposal.md

```diff
@@ -0,0 +1,161 @@
+    size_t       sealed_blob_size;            
+    uint32_t     reserved2[2];                /* Must be 0 */
+	/* For security, sealed blob is copied into enclave */
+    uint8_t      sealed_blob[PCL_SEALED_BLOB_SIZE];
+	/* SHA256 digest of decryption key */
+    uint8_t      decryption_key_hash[SGX_SHA256_HASH_SIZE];
+	/* Number of RVAs */
+    uint32_t     num_rvas;                    
+    uint32_t     reserved3[3];                /* Must be 0 */
+	/* Array of rva_size_tag_iv_t */
+    rva_size_tag_iv_t rvas_sizes_tags_ivs[PCL_MAX_NUM_ENCRYPTED_SECTIONS]; 
+}pcl_table_t;
+
+### Sections left plaintext
+1. ELF header  - binary header
+2. Sections table
+3. Segments table
+4. Sections' names string table pointed by e_shstrndx(e.g. .shstrtab)
+5. .oeinfo section holds enclave's metadata(properties)
+6. .bss and .tbss
+7. sections required to construct dyn_info (.dynamic)
+8. sections holds the content provided by entries with index DT_SYMTAB, DT_STRTAB and DT_REL in
+   dyn_info (e.g. .dynsym, .dynstr, .rela.dyn)
+9. sections containing PCL code and data:
+   a. section ".pcltbl"  // Designated section for PCL table
+   b. .nipx, .nipd, .niprod, .nipd_rel, .nipd_rel_ro_local
+10. sections for debugging
+   .comment, .debug_abbrev, .debug_aranges, .debug_info, .debug_line, .debug_lc, .debug_ranges,
+   .debug_str
+
+## Elf sections to be encrypted
+Sections not mentioned above are sections containing IP information that need to be protected.
+Mainly those sections are:
+1. Code sections -- .text in elf file
+2. Data sections -- .data in elf file (initialized local/global variables)
+3. Read-only Data sections -- .rodata in elf file (const variables)
+4. sections containing relocation info related with the above items
+
+## Sealing/Unsealing the decryption key
+To deliver the decryption key in a secure way, the key used to encrypt the enclave should get sealed
+by key policy OE_SEAL_POLICY_PRODUCT on deliverting to Protected Code Loader for decryption in a 
+sealed_blob. The protected code loader will unseal this sealed_blob to get the decryption key. 
+
+## new APIs and libs
+### Encryption tool
+new tool for ELF image encryption - oeencrypt: placed in OESDK installation folder as the bin files
+new lib for section ".pcltbl" and decryption -- liboepcl.a: placed in OESDK installation folderd
+as the lib files
+
+### Modifications to OE SDK runtime lib
+No new APIs exposed to user.
+1. A new setting is defined in include/openenclave/host.h as the argument to support
+protected code loader enclave loading in API oe_create_enclave.
+typedef enum _oe_enclave_setting_type
+{
+	// PCL setting type
+	OE_ENCLAVE_SETTING_PCL = 0xac120002,
+	...
+} oe_enclave_setting_type_t;
+
+typedef struct _oe_enclave_setting
+{
+    oe_enclave_setting_type_t setting_type;
+    /**
+     * The specific setting for the enclave, such as for configuring
+     * context-switchless calls.
+     */
+    union {
+		/* for protected code loader */
+		uint8_t* sealed_blob;
```

Comment:
Where does the app get this blob from?   Add answer in the doc somewhere.

Reply:
Assuming the sealed_blob isn't the actual code to be decrypted, this seems to make an assumption that the enclave has 1 and only 1 possibility of what to decrypt.  I'd prefer a solution that does not force that assumption.
The encrypted code may be acquired at runtime (e.g., across a network), or there may be multiple possible ones compiled into the enclave, or whatever else.    Things get easier if the granularity is at the module level and not specific pages or segments, and it becomes more generally applicable across TEEs.

Reply:
For use who's using PCL feature, they will not be aware of the encryption/decryption function of their enclave image - ELF file, which specific sections need to be encrypted/decrypted are done by pcl in oe sdk. The solution proposed here, we provided a simple way for user to use PCL, they don't need to do anything regarding the encryption/decryption algorithm, what they need to do is only to follow the PCL sample code to provide a raw enclave image, linking it to oepcl lib. Encryption is by oeencrypt, and decrypt is by oe sdk runtime.

Reply:
Can you make the work flow more clear in this document? How do user to generate a sealed blob is not clear in this document.

Reply:
I second @mingweishih. The purpose of sealed blob is not clear.

Reply:
thanks, I will update a seperate title for the sealed_blob in PCL entry. The purpose of the sealed_blob is based on the thought of delivering the decryption key in a secure way.